### PR TITLE
initial implementation

### PR DESCRIPTION
**Pull request by rogpeppe**
This implements the environment schema as described in https://docs.google.com/document/d/1vfQd9ZAKOGysqMe0AICecgejky85Us5ro5aH0PpejGo.

### CODE REVIEW

File: fields.go [hunk 1]
```diff
@@ -0,0 +1,156 @@
+// Package environschema implements a way to specify
```
Comment (mhilton) [inline fields.go:4]: Copyright notice?

File: fields.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+// configuration attributes for Juju environments.
+package environschema // gopkg.in/juju/environschema.v1
+import (
```
Comment (mhilton) [inline fields.go:8]: isn't there normally a blank line between package and import

File: fields.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
+// Package environschema implements a way to specify
+// configuration attributes for Juju environments.
+package environschema // gopkg.in/juju/environschema.v1
```
Comment (mhilton) [inline fields.go]: // import "gopkg.in/juju/environschema.v1" ?

File: fields.go [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+	if len(path) == 0 {
+		return ""
+	}
+	if path[0] == "." {
+		return strings.Join(path[1:], "") + ": "
+	}
+	return strings.Join(path, "") + ": "
+}
+
+// func (s Fields) Checker() schema.Checker
```
Comment (mhilton) [inline fields.go]: What are these for, things to be added?

File: package_test.go [hunk 5]
```diff
@@ -0,0 +1,14 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
```
Comment (mhilton) [inline package_test.go]: LICENSE file says it's LGPL

**Comment by mhilton** [comment]
:+1: LGTM with a couple of comments

File: fields.go [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+	// Values holds the set of all possible values of the attribute.
+	Values []interface{} `json:"values,omitempty"`
+}
+
+// FieldType describes the type of an attribute value.
+type FieldType string
+
+// The following constants are the possible type values.
+const (
+	Tstring FieldType = "string"
```
Comment (urosj) [inline fields.go:63]: FTstring maybe? (also FTbool and FTint) ... although Tstring is ok as well.

File: fields.go (hunk 1 above)
Comment (rogpeppe) [inline fields.go:4]: done

File: fields.go (hunk 2 above)
Comment (rogpeppe) [inline fields.go:8]: done

File: fields.go (hunk 3 above)
Comment (rogpeppe) [inline fields.go]: done

File: package_test.go (hunk 5 above)
Comment (rogpeppe) [inline package_test.go]: done

File: fields.go (hunk 4 above)
Comment (rogpeppe) [inline fields.go]: Possible alternatives. I've moved the comment and made it a bit clearer.

File: fields.go (hunk 6 above)
Comment (rogpeppe) [inline fields.go:63]: I'd be more inclined to rename FieldType to Type than to use FT, I think, if you want that to be consistent. I'm reasonably happy with the T prefix, although I'm not entirely sure that it shouldn't be TString, TBool, etc.

File: fields.go [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+		checker: checker,
+	}, nil
+}
+
+type oneOfValuesC struct {
+	vals    []interface{}
+	checker schema.Checker
+}
+
+func (c oneOfValuesC) Coerce(v interface{}, path []string) (interface{}, error) {
```
Comment (urosj) [inline fields.go:138]: Please add method description.

File: fields.go [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+	Tint    FieldType = "int"
+)
+
+var checkers = map[FieldType]schema.Checker{
+	Tstring: schema.String(),
+	Tbool:   schema.Bool(),
+	Tint:    schema.ForceInt(),
+}
+
+func (s Fields) ValidationSchema() (schema.Fields, schema.Defaults, error) {
```
Comment (urosj) [inline fields.go:82]: Please add description.

File: fields.go (hunk 4 above)
Comment (urosj) [inline fields.go]: +1 on the question

File: fields_test.go [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+	fields: environschema.Fields{
+		"stringvalue": {
+			Type: environschema.Tstring,
+		},
+		"mandatory-stringvalue": {
+			Type:      environschema.Tstring,
+			Mandatory: true,
+		},
+		"intvalue": {
+			Type: environschema.Tint,
```
Comment (urosj) [inline fields_test.go:41]: What about testing for secret? Maybe a bit too much but can bool/int be mandatory?

**Comment by urosj** [comment]
:+1: Ty for this. Nice start.

File: fields.go (hunk 8 above)
Comment (rogpeppe) [inline fields.go:82]: done

File: fields_test.go (hunk 9 above)
Comment (rogpeppe) [inline fields_test.go:41]: they can, but it doesn't test any more code paths.

**Pull request merged**